java-checks/src/main/java/org/sonar/java/checks/ObjectFinalizeOverridenCallsSuperFinalizeCheck.java

```diff
@@ -47,76 +47,91 @@ public class ObjectFinalizeOverridenCallsSuperFinalizeCheck extends IssuableSubs
+      Type superClassType = ((Symbol.TypeSymbol) classSymbol).superClass();
+      while (superClassType != null && !superClassType.is("java.lang.Object")) {
+        final Symbol.TypeSymbol currentClass = superClassType.symbol();
+        if (currentClass.lookupSymbols(FINALIZE).stream().anyMatch(FINALIZE_MATCHER::matches)) {
+          return true;
         }
+        superClassType = currentClass.superClass();
       }
     }
+    return false;
   }
 
-  private static boolean isLastStatement(MethodTree methodTree, MethodInvocationTree lastStatementTree) {
-    BlockTree blockTree = methodTree.block();
-    if (blockTree != null
-      && blockTree.body().stream().anyMatch(statement -> 
-        statement.is(Tree.Kind.TRY_STATEMENT)&& isLastStatement(((TryStatementTree) statement).finallyBlock(), lastStatementTree))) {
+  @Nullable
+  private static MethodInvocationTree findLastSuperFinalizeInvocation(BlockTree blockTree) {
+    final FindLastSuperFinalizeInvocationVisitor visitor = new FindLastSuperFinalizeInvocationVisitor();
+    blockTree.accept(visitor);
+    return visitor.getLastSuperFinalizeInvocation();
+  }
+
+  private static boolean isLastStatement(BlockTree blockTree, MethodInvocationTree lastStatementTree) {
+    if (blockTree.body().stream().anyMatch(statement ->
+      statement.is(Tree.Kind.TRY_STATEMENT) && isLastStatementInner(((TryStatementTree) statement).finallyBlock(), lastStatementTree)
+    )) {
       return true;
     }
-    return isLastStatement(blockTree, lastStatementTree);
+    return isLastStatementInner(blockTree, lastStatementTree);
   }
 
-  private static boolean isLastStatement(@Nullable BlockTree blockTree, MethodInvocationTree lastStatementTree) {
+  private static boolean isLastStatementInner(@Nullable BlockTree blockTree, MethodInvocationTree lastStatementTree) {
     if (blockTree != null) {
-      StatementTree last = ListUtils.getLast(blockTree.body());
+      final StatementTree last = ListUtils.getLast(blockTree.body());
       if (last.is(Tree.Kind.EXPRESSION_STATEMENT)) {
         return lastStatementTree.equals(((ExpressionStatementTree) last).expression());
       } else if (last.is(Tree.Kind.TRY_STATEMENT)) {
-        return isLastStatement(((TryStatementTree) last).finallyBlock(), lastStatementTree);
+        return isLastStatementInner(((TryStatementTree) last).finallyBlock(), lastStatementTree);
       }
     }
     return false;
   }
 
-  private static boolean doesOverrideFinalize(Symbol classSymbol) {
-    if (classSymbol.isTypeSymbol()) {
-      Type superClassType = ((Symbol.TypeSymbol) classSymbol).superClass();
-      while (superClassType != null && !superClassType.is("java.lang.Object")) {
-        Symbol.TypeSymbol currentClass = superClassType.symbol();
-        if (currentClass.lookupSymbols(FINALIZE).stream().anyMatch(FINALIZE_MATCHER::matches)) {
-          return true;
+  private static class FindLastSuperFinalizeInvocationVisitor extends BaseTreeVisitor {
```

Comment:
I believe this is not enough. You need to avoid exploring inner classes or anonymous classes in your base tree visitor. The following case will trigger both FNs and FPs:

```Java
class FpAndFn extends S1114_Class1 {
  @Override
  protected void finalize() throws Throwable {  // False Negative: there is no call to finalize here
    class InnerClass {
      void foo() throws Throwable {
        super.finalize(); // False positive: this call is NOT inside a finalize method();
      }
    }
  }
}
```